Parse full information for each sensor

chenj47 opened the pull request:
Three key changes are made in this pull request:
1. Fix the bug in the old battery historian tool where if the dump section for sensorservice starts with "DUMP OF SERVICE CRITICAL sensorservice:", then the information cannot be obtained. Now the tag CRITICAL will be handled.
2. Rename the battery-historian file to sensor-historian so that to run the tool now, the command line is "go run cmd/sensor-historian/sensor-historian.go" which corresponds to the instruction in README file.
3. Change the data recorded by data structure bugreportutils.SensorInfo and the related function extractSensorInfo so that full information about each sensor will be obtained. New test cases are added for this update.

chenj47 committed:
Fix a bug in old Battery Historian repo + Rename the battery-historian file to sensor-historian + Get full information for each sensor

agarwal-v commented on presenter/presenter.go:
```diff
@@ -48,20 +48,20 @@ func absInt32(x int32) int32 {
 }
 
-// userActivity contains a processed form of the UserActivity proto found in github.com/googleinterns/sensor-historian/pb/batterystats.proto.
+// userActivity contains a processed form of the UserActivity proto found in github.com/google/sensor-historian/pb/batterystats.proto.
```
googleinterns

agarwal-v replied:
interns

agarwal-v commented on presenter/presenter.go:
```diff
@@ -48,20 +48,20 @@ func absInt32(x int32) int32 {
 
 // AppStat contains the parsed app data from a bugreport.
-// This contains the raw App proto in github.com/googleinterns/sensor-historian/pb/batterystats.proto
+// This contains the raw App proto in github.com/google/sensor-historian/pb/batterystats.proto
```
google interns

agarwal-v replied:
interns

agarwal-v commented on presenter/presenter.go:
```diff
@@ -248,7 +248,7 @@ func (a byName) Len() int           { return len(a) }
 func (a byName) Less(i, j int) bool { return a[i].RawStats.GetName() < a[j].RawStats.GetName() }
 
-func parseAppStats(checkin *bspb.BatteryStats, sensors map[int32]bugreportutils.SensorInfo) []AppStat {
+func parseAppStats(checkin *bspb.BatteryStats, sensors map[int32]*bugreportutils.SensorInfo) []AppStat {
```
Why change the function signature?

agarwal-v commented on historianutils/historianutils.go:
```diff
@@ -31,6 +31,9 @@ var (
 	ServiceDumpRE = regexp.MustCompile(`^DUMP\s+OF\s+SERVICE\s+(?P<service>\S+):`)
 
+	// ServiceDumpWithTagRE is a regular expression to match the beginning of a service dump with tag.
+	ServiceDumpWithTagRE = regexp.MustCompile(`^DUMP\s+OF\s+SERVICE\s+(?P<serviceTag>\S+)\s+(?P<service>\S+):`)
```
Is the service tag useful? If not is it possible to created just a single regular expression that works with or without service tag?

agarwal-v commented on bugreportutils/bugreportutils.go:
```diff
@@ -128,15 +144,22 @@ type MetaInfo struct {
+	Max, Reserved int32
+	WakeUp        bool
+	TotalTimeMs   int64 // time.Duration in Golang is converted to nanoseconds in JS, so using int64 and naming convention to be clear in$
+	Count         float32
```
Can we keep the first few lines same as the previous version?

agarwal-v commented on bugreportutils/bugreportutils.go:
```diff
@@ -128,15 +144,22 @@ type MetaInfo struct {
 	BuildFingerprint string
 	ModelName        string
-	Sensors          map[int32]SensorInfo
+	Sensors          map[int32]*SensorInfo
```
I'd consider if avoiding this change makes sense.

chenj47 replied:
Currently I have a map[int32]*SensorInfo, this allows me to directly modify the struct that is referenced by the pointer. So suppose I am updating the name of sensor with handle 1, I can just do:
```
sensor := map[1]
sensor.Name = "newName"
```
However with the old map[int32]SensorInfo type, to accomplish the same thing, I need to:
```
sensor := map[1]
sensor.Name = "newName"
map[1] = sensor
```
Since accessing the sensorInfor in the map gives me a copy of the value, and I need to overwrite the copy again.

agarwal-v replied:
But you can always use map[1].Name without any impact, I think... & where is the code you are using that is impacted in this change.

I'm worried about this as when we start merging this back, similar questions might be asked.

chenj47 replied:
In golang, this following line does not work:
```
map[1].Name := "newName"
```
The compiler will throw a panic warning to this. I will stick with the pointer struct for now.

chenj47 replied:
I totally agree that the same question might be asked again when we start merging this back. Also considering that this change may impact other files, I will change it back to the original struct rather than using the pointer in the map.

agarwal-v reviewed (commented):
Minor comments + a few questions.

agarwal-v commented on bugreportutils/bugreportutils.go:
```diff
@@ -43,35 +43,51 @@ var (
+	// sensorLineOneRE is a regular expression to match the first line of sensor information
+	// from the sensor list line in the sensorservice dump in the bugreport starting
+	// from NRD42 and onwards.
+	sensorLineOneRE = regexp.MustCompile(`(?P<sensorHandle>0x?[0-9A-Fa-f]+)` + `\)\s*` +
```
Why change sensorNumber to sensorHandle?

ALso deleting sensorLineMMinus will break backward compatibility.

chenj47 replied:
I was use both "handle" and "number" to refer to the unique identification number for each sensor. To keep things consistent, I will only be using the word "number" from now on.

Regarding the backward compatibility, I will need to look at more examples for older bug report to determine how to parse the information.

chenj47 committed:
Fix the comments in the presenter.go file + Use one expression to match the DUMP OF SERVICE line and check for substring "sensorservice"

chenj47 committed:
Stick with the old SensorInfo struct in the map instead of the pointer

rglasser commented on bugreportutils/bugreportutils.go:
```diff
@@ -137,6 +153,11 @@ type SensorInfo struct {
 	TotalTimeMs     int64 // time.Duration in Golang is converted to nanoseconds in JS, so using int64 and naming convention to be clear in$
 	Count           float32
+	RequestMode     string
+	Var1, Var2      string
```
Can we use more descriptive variable names than Var1 and Var2?

chenj47 replied:
Yes, I will just use MaxRate, MinRate, etc to store these values.

agarwal-v commented on historianutils/historianutils.go:
```diff
@@ -29,7 +29,7 @@ import (
 var (
 	// ServiceDumpRE is a regular expression to match the beginning of a service dump.
-	ServiceDumpRE = regexp.MustCompile(`^DUMP\s+OF\s+SERVICE\s+(?P<service>\S+):`)
+	ServiceDumpRE = regexp.MustCompile(`^DUMP\s+OF\s+SERVICE\s+(?P<service>.*):`)
```
Can you add a comment that the tag <service> will include the optional priority of the service like HIGH or CRITICAL along with service name

chenj47 replied:
Sure, I will include that in the next commit.

chenj47 committed:
Added Comments + Accomodate the sensor list format for Android version from MNC or before

chenj47 committed:
Add the testcase to accommodate sensor Number being 00000000

chenj47 committed:
Parse the minRate/maxRate or maxDelay/minDelay values and store the delay in us rather than rate in Hz. Also accommodate this feature to older Android Version

chenj47 committed:
Fix spacing

agarwal-v commented on bugreportutils/bugreportutils.go:
```diff
@@ -43,35 +43,63 @@ var (
+	BatchingDataRE = regexp.MustCompile(`FIFO\s*\(max,reserved\)\s*=\s*\(` + `(?P<maxNum>\d+)` +
+		`,` + `(?P<reservedNum>[^|]+)` + `\) events\s*`)
+
+	rateRE    = regexp.MustCompile(`=(?P<rateVal>[\-\+]?[0-9]*(\.[0-9]+)+)Hz`)
```
A comment stating that rateRE, delayRE & fifoMAxRE are to parse the sensor information from the dumpsys.

agarwal-v commented on bugreportutils/bugreportutils.go:
```diff
@@ -133,10 +161,15 @@ type MetaInfo struct {
-	Count           float32
+	Name, Type         string
+	Version, Number    int32
+	TotalTimeMs        int64 // time.Duration in Golang is converted to nanoseconds in JS, so using int64 and naming convention to be clear in$
```
Wrap the comment to fir in 80 characters....

agarwal-v commented on historianutils/historianutils.go:
```diff
@@ -105,6 +108,14 @@ func MaxInt64(a int64, b int64) int64 {
 }
 
+// RoundFloat returns the closest integer to the given float number.
+func RoundFloat(val float64) int32 {
```
Why not use https://golang.org/pkg/math/#Round
math.Round()

chenj47 replied:
I am not using that function because its return type is still float64, while in my case I want to use int32. I will rename this function to RoundFloatToInt and make use of the math.Round() function inside.

agarwal-v reviewed (commented):
Minor comments!!

chenj47 committed:
Add comments and fix spacing in the bugreportutils/ and historianutils/ folders
Add more comments to explain the regular expressions used.
Fix the spacing in some of the functions so that each chunk of code is doing one thing.
Reformat some comments so that most of the lines are under 100 characters. Note that the traditional 80 characters limit is not adopted since many lines of code are calling functions from other file so the long package name is unevitable. Some lines that involve if statements are not strictly under 100 characters because of Golang convention.
